src/utils/allocators.zig

```diff
@@ -0,0 +1,114 @@
+const std = @import("std");
+
+pub const RecycleFBA = struct {
+    // this allocates the underlying memory + dynamic expansions
+    // (only used on init/deinit + arraylist expansion)
+    backing_allocator: std.mem.Allocator,
+    // this does the data allocations (data is returned from alloc)
+    alloc_allocator: std.heap.FixedBufferAllocator,
+    // recycling depot
+    records: std.ArrayList(Record),
+
+    // for thread safety
+    mux: std.Thread.Mutex = .{},
+
+    const Record = struct { is_free: bool, buf: [*]u8, len: u64, log2_align: u8 };
```

Comment:
We don't need to store the alignment as a value, we can simply check whether the pointer is aligned with `std.mem.isAlignedLog2(@intFromPtr(ptr), log2_align)`.

Reply:
ooo cool 170c652